File: src/www/__test__/testPost/reader/PostSection.ts
import { PostContent } from "./PostContent";
import { PostLine } from "./PostLine";
import { PostSubsection } from "./PostSubsection";

export class PostSection {
  private subsectionsList: PostSubsection[] = [];
  private subsectionsByName: { [name: string]: PostSubsection } = {};

  constructor(private content: PostContent, private sectionName: string) {}

  addLine(subsectionName: string, line: PostLine) {
    if (!this.subsectionsByName[subsectionName]) {
      const newSubsection = new PostSubsection(subsectionName);
      this.subsectionsByName[subsectionName] = newSubsection;
      this.subsectionsList.push(newSubsection);
    }

    const subsection = this.subsectionsByName[subsectionName];
    subsection.addLine(line);
  }

  getLines(): PostLine[] {
    return this.subsectionsList.flatMap((ss) => ss.getLines());
  }

  getSectionName(): string {
    return this.sectionName;
  }

  getPrettyName(): string {
    const firstLine = this.getFirstLine();
    return `${firstLine.getContent()} (${this.content.getPostId()})`;
  }

  private getFirstLine(): PostLine {
    const firstSubsection = this.subsectionsList[0];
    const line = firstSubsection.getFirstLine();
    return line;
  }
}

File: src/test/java/com/drpicox/game/testPost/reader/PostSection.java
package com.drpicox.game.testPost.reader;

import java.util.*;

public class PostSection {
    private final Map<String, PostSubsection> subsections = new LinkedHashMap<>();
    private final PostContent content;
    private final String sectionName;

    public PostSection(PostContent content, String sectionName) {
        this.content = content;
        this.sectionName = sectionName;
    }

    public void addLine(String subsectionName, PostLine line) {
        if (!subsections.containsKey(subsectionName))
            subsections.put(subsectionName, new PostSubsection(subsectionName));

        var subsection = subsections.get(subsectionName);
        subsection.addLine(line);
    }

    public List<PostLine> getLines() {
        var lines = new ArrayList<PostLine>();
        for (var subsection: subsections.values())
            lines.addAll(subsection.getLines());

        return lines;
    }

    public String getSectionName() {
        return sectionName;
    }

    public String getPrettyName() {
        var firstLine = getFirstLine();
        return firstLine.getContent() + " (" + content.getPostId() + ")";
    }

    private PostLine getFirstLine() {
        var firstSubsection = subsections.values().stream().findFirst().get();
        var line = firstSubsection.getFirstLine();
        return line;
    }

}
